### PR TITLE
Refactor and allow to filter in networks with no TCR

### PR DESCRIPTION
**Pull request by anxolin**
Part of https://github.com/gnosis/dex-react/issues/1398

This PR was created to avoid having to deploy a TCR for xDAI. 

The current behaviour of the app, is that **if there's no TCR or it's list is empty, then it doesn't apply any filters, so it will display ALL THE TOKENS**

This behaviour is a bit undesirable. For xDAI we have a default list of some basic tokens. Anyone deploying the app can deploy the TCR for xDAI but if it's not done, it will use the default list.

Also, moves the part of getting the ids and filtering them to its own function. The logic was already pretty complex, and was using too many loops. 

**NOT PART OF THIS PR**
The token loading is unnecessarily complex. This PR doesn't try to tackle that, although it does a small refactor on some parts it touches, the goal is not to do a full rewrite of this. It would be a good idea to invest in cleaning up this.

See: https://github.com/gnosis/dex-react/pull/1489#discussion_r500142151

### CODE REVIEW

**Comment by gnosis-info** [comment]
Travis automatic deployment:
   * **🎩 [Dapp](https://pr1489--dexreact.review.gnosisdev.com)**: Testing web app
   * **📚 [Storybook](https://pr1489--dexreact.review.gnosisdev.com/storybook/)**: Component stories

File: src/services/factories/tokenList.ts [hunk 1]
```diff
@@ -130,42 +130,57 @@ export function getTokensFactory(factoryParams: {
...
+      logDebug(`[tokenListFactory][${networkId}] Not using a TCR. Filtering by tokens in the config file`)
+      const tokensConfig = tokenListApi.getTokens(networkId)
+      const filteredAddressAndIds = tokensConfig
+        .map((token): [string, number] | null => {
+          const tokenId = listedAddressesAndIds[token.address]
+          return [token.address, tokenId] || null
+        })
+        .filter(notEmpty)
+
+      return new Map<string, number>(filteredAddressAndIds)
```
Comment (W3stside) [inline src/services/factories/tokenList.ts:162]: you don't want to move this outside the `if/else` with a `let filteredAddressAndIds` outside the scope? seeing as you're returning `new Map(filteredAddressAndIds` in both cases

File: src/services/factories/tokenList.ts [hunk 2]
```diff
@@ -191,7 +206,7 @@ export function getTokensFactory(factoryParams: {
           `[tokenListFactory][${networkId}] Got details for address ${partialToken.address}: symbol '${partialToken.symbol}' name '${partialToken.name}'`,
         )
         // Get id from address/id mapping
-        const id = addressesAndIds.get(partialToken.address) as number
+        const id = filteredAddressesAndIds.get(partialToken.address) as number
```
Comment (W3stside) [inline src/services/factories/tokenList.ts:209]: can't this be `undefined`?

**Review by W3stside — APPROVED** [review]

File: src/services/factories/tokenList.ts (hunk 1 above)
Comment (anxolin) [reply src/services/factories/tokenList.ts:162]: I considered it, but just went for this option. I'm ok change it too. In this case I don't feel makes a huge difference

Comment (anxolin) [reply src/services/factories/tokenList.ts]: 5b8e19c3e83925d232bdfb2ac7dcbab98bb0b40f

File: src/services/factories/tokenList.ts (hunk 2 above)
Comment (anxolin) [reply src/services/factories/tokenList.ts:209]: I don't think so. The logic for loading the tokens is a bit unnecessarily complex. This is why these things are required.
I did one small refactor in this PR, but they could have more refactors and probably this `as number` would disapear. But this is not part of this PR refactor all.

It cannot happen to be undefined, cause we made sure in the code above that only tokens listed in the exchange (with id) are in this `partialTokens` array

File: src/services/factories/tokenList.ts [hunk 3]
```diff
@@ -130,42 +130,57 @@ export function getTokensFactory(factoryParams: {
...
     )
 
     const promises: Promise<TokenErc20 | string>[] = []
 
     // Go over all value (id) and key (tokenAddress) pairs registered on the contract
-    addressesAndIds.forEach((id, tokenAddress) => {
+    filteredAddressesAndIds.forEach((id, tokenAddress) => {
       if (!localAddressesMap.has(tokenAddress)) {
         // New token not in our local list, fetch erc20 details for it
         logDebug(`[tokenListFactory][${networkId}] Address ${tokenAddress} with id ${id} not in local list, fetching`)
```
Comment (Velenir) [inline src/services/factories/tokenList.ts:186]: This looks complicated

Is the idea to:

1. get intersection of tokens from `tokenListApi.getTokens(networkId)` and `filteredAddressesAndIds`
2. fetch tokens that are in `filteredAddressesAndIds` but not in `tokenListApi.getTokens(networkId)` as ERC20
3. fill in the blanks of tokens from 2 from chain as ERC20
4. assign ids to fetched ERC20

If I'm correct, then I feel you have too many loops and a pretty complicated logic in general. Would the following do the same?

```ts
async function updateTokenDetails(networkId: number, numTokens: number): Promise<void> {
  // Get filtered ids and addresses
  const filteredAddressesAndIds = await _getFilteredIdsMap(networkId, numTokens)
  const localTokens: TokenDetails[] = []

  const allTokensFromApi = new Map(tokenListApi.getTokens(networkId).map((t) => [t.address, t]))
  const TokenDetailsPromises: Promise<TokenDetails | undefined>[] = []

  filteredAddressesAndIds.forEach((id, tokenAddress) => {
    const token = allTokensFromApi.get(tokenAddress)
    if (token) {
      localTokens.push(token)
    } else {
      const fullTokenPromise = getErc20DetailsOrAddress(networkId, tokenAddress).then((partialToken) => {
        if (typeof partialToken === 'string') {
          // We replaced potential null responses with original tokenAddress string for logging purposes
          logDebug(`[tokenListFactory][${networkId}] Address ${partialToken} is not a valid ERC20 token`)
          return
        }
        // If we got a valid response
        logDebug(
          `[tokenListFactory][${networkId}] Got details for address ${partialToken.address}: symbol '${partialToken.symbol}' name '${partialToken.name}'`,
        )
        // build token object
        const token: TokenDetails = { ...partialToken, id }

        return token
      })
      TokenDetailsPromises.push(fullTokenPromise)
    }
  })

  const fetchedAndFilledTokenDetails = (await Promise.all(TokenDetailsPromises)).filter(notEmpty)

  const tokenList = localTokens.concat(fetchedAndFilledTokenDetails)

  // Persist it \o/
  tokenListApi.persistTokens({ networkId, tokenList })
}
```

**Review by Velenir — CHANGES_REQUESTED** [review]
This won't work, sorry

File: src/services/factories/tokenList.ts [hunk 4]
```diff
@@ -130,42 +130,57 @@ export function getTokensFactory(factoryParams: {
...
 
-    logDebug(`[tokenListFactory][${networkId}] TCR contains ${tcrAddressesSet.size} addresses`)
+    if (tcrAddressesSet.size > 0) {
+      // If filtering by TCR
+      logDebug(`[tokenListFactory][${networkId}] TCR contains ${tcrAddressesSet.size} addresses`)
+      logDebug(`[tokenListFactory][${networkId}] Token id and address mapping:`)
+      const filteredAddressAndIds = Array.from(listedAddressesAndIds.entries()).filter(([address, id]) => {
+        // Remove addresses that are not on the TCR, if any
+        const isInTcr = tcrAddressesSet.has(address)
+        logDebug(`[tokenListFactory][${networkId}] ${id} : ${address}. On TCR? ${isInTcr}`)
```
[the remark this answers is not in the document]
Comment (anxolin) [reply src/services/factories/tokenList.ts:149]: Great catch! I think I rushed this PR too much!

File: src/services/factories/tokenList.ts [hunk 5]
```diff
@@ -130,42 +130,57 @@ export function getTokensFactory(factoryParams: {
...
+      })
+      return new Map<string, number>(filteredAddressAndIds)
+    } else {
+      // If not using a TCR, filter by default list
+      logDebug(`[tokenListFactory][${networkId}] Not using a TCR. Filtering by tokens in the config file`)
+      const tokensConfig = tokenListApi.getTokens(networkId)
+      const filteredAddressAndIds = tokensConfig
+        .map((token): [string, number] | null => {
+          const tokenId = listedAddressesAndIds[token.address]
+          return [token.address, tokenId] || null
```
[the remark this answers is not in the document]
Comment (anxolin) [reply src/services/factories/tokenList.ts:159]: Yep, this is derived from the previous error on how I was accessing the Map in a Object fashion

File: src/services/factories/tokenList.ts (hunk 3 above)
Comment (Velenir) [reply src/services/factories/tokenList.ts:187]: @anxolin did you consider this refactor?

Comment (W3stside) [reply src/services/factories/tokenList.ts:187]: :point_up: @Velenir's logic looks correct and is easier to read imo

Comment (anxolin) [reply src/services/factories/tokenList.ts:187]: Proposal that reiterate on Dima's ideas. I think it simplifies it a bit further and fix some weird signature we had in the API
https://github.com/gnosis/dex-react/pull/1498/files

**Review by Velenir — APPROVED** [review]

**Pull request merged**